app/controllers/vouchers_controller.rb

```diff
@@ -58,17 +58,49 @@ def create
       @existing_voucher = @swap.vouchers.find_by(client_id: @voucher.client_id)
       return render :new
     end
-
+    
     if !@voucher.save
       return render :new
     end
 
+
     redirect_to action: :created, id: @voucher.id
   end
 
   def created
   end
 
+  def send_voucher
+    # EMAIL
+    if params[:commit] == 'E-mail'
+      @voucher = Voucher.find(params[:id])  
+      
+      VoucherMailer.with(voucher: @voucher).voucher_email.deliver_now
+     
+
+    # TEXT/SMS
+    elsif params[:commit] == 'Text/SMS'
+      # CLIENT VOUCHER
+      @voucher = Voucher.find(params[:id])
+
+      # CREDS
+      account_sid = ENV['TWILIO_SID']
+      auth_token = ENV['TWILIO_TOKEN']
+
+      # MESSAGE BODY
+      digi_voucher = "Thanks, #{@voucher.client.name}! Your voucher for #{@voucher.hotel.name} is confirmed!\n\nVoucher #: #{@voucher.number}\n\n#{@voucher.hotel.name} is expecting you from #{@voucher.check_in} to #{@voucher.check_out}.\n\nAddress:\n\n#{@voucher.hotel.street_address}\n#{@voucher.hotel.address_second}\n#{@voucher.hotel.phone}"
+      
+      # API CALL
+      client = Twilio::REST::Client.new(account_sid, auth_token)
+      client.messages.create(
+        from: ENV['TWILIO_NUMBER'],
+        to: @voucher.client.phone_number,
+        body: digi_voucher
+      )
```

Comment:
The Twilio code seems to occur on a different level of encapsulation than the email. If we're following SOLID principles, this controller does one thing: receive routed messages, setup context needed for the response, and then let another layer of the system handle the generation of the response.

In "regular" controller actions, the Rails view layer is handling the actual composing of the html response. Here, we're using multiple paradigms -- the email path hands off the responsibility of creating the response to the VoucherMailer (which uses a view-like layer for the email templating), but in the sms path, we do it inline.

Can we make both paths consistent in their level of encapsulation? If they were, I would expect this controller action to look something like...
```
if params[:commit] == "Email"
  VoucherMailer.with(voucher: @voucher).voucher_email.deliver_now
elsif params[:commit] == "Text/SMS"
  TwilioHandler.send_voucher_sms(voucher: @voucher) # or whatever API makes sense
end
```

^ that keeps the controller doing only one thing: routing the request data to the right handler.

Reply:
I'll need more guidance on this one. Reading what you sent me...I do not understand, even with the architecture explanation (do text messages need a full view?). Checking the files in `app/helpers`, my helper looks the same, and I get an error that my method does not exist. Rebooted spring and console as well.